src/middleware/open-graph/index.js

```diff
@@ -0,0 +1,165 @@
+
+const fetchStats = async (name, type = 'npm', period = 'month') => {
+	return http.get(`${API_HOST}/v1/package/${type}/${name}/stats/date/${period}`).json();
+};
+
+const fetchLogo = async (url) => {
+	return new Promise((resolve, reject) => {
+		let stream = http.stream.get(url);
+
+		let size = 0;
+		let mime = 'image/png';
+		let logo = [];
+
+		stream.on('end', () => {
+			resolve(`data:${mime};base64,${Buffer.concat(logo).toString('base64')}`);
+		});
+
+		stream.on('error', reject);
+
+		stream.on('response', (res) => {
+			if (res.headers['content-length'] && res.headers['content-length'] > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			mime = res.headers['content-type'] ?? 'image/png';
+		});
+
+		stream.on('data', (chunk) => {
+			size += chunk.length;
+
+			if (size > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			logo.push(chunk);
+		});
+	});
+};
+
+const processDescription = (description) => {
+	let charsPerLine = 65;
+	let lineOffset = 205;
+	let lineHeight = 36;
+
+	let lines = wordwrap(charsPerLine)(description.substring(0, charsPerLine * 2)).split('\n');
+
+	if (lines.length > 2) {
+		lines = lines.slice(0, 2);
+		let lastLine = lines.pop();
+
+		if (lastLine.length > charsPerLine - 3) {
+			lastLine = lastLine.split(' ').slice(0, -1).join(' ');
+		}
+
+		lines.push(lastLine + '...');
+	}
+
+	lines = lines.map((text, idx) => {
+		return { text, offset: lineOffset + lineHeight * idx };
+	});
+
+	return lines;
+};
+
+const processRequestsChart = (records, max) => {
+	let barHeight = 40;
+	let barPadding = 10;
+	let offsetX = 88;
+	let offsetY = 418;
+
+	return records.map((record, idx) => {
+		let h = Math.floor((record.total / max) * barHeight);
+		let x = offsetX + barPadding * idx;
+		let y = offsetY + barHeight - h;
+
+		return { h, x, y };
+	});
+};
+
+const prepareMetadata = async (pkg) => {
+	let cacheKey = `algolia::${pkg}`;
+
+	if (cache.has(cacheKey)) {
+		return cache.get(cacheKey);
+	}
+
+	let meta = await npmIndex.getObject(pkg, [ 'name', 'description', 'version', 'owner' ]);
+
+	meta.owner.logo = await fetchLogo(meta.owner.avatar).catch(() => {});
+	meta.description = processDescription(meta.description);
+
+	cache.set(cacheKey, meta);
+
+	return meta;
+};
+
+const prepareStats = async (name) => {
+	let stats = await fetchStats(name);
+
+	let formatNum = num => num.toLocaleString().replaceAll(',', ' ');
+	let max = 0;
+
+	let records = Object.values(stats.dates).map((stats) => {
+		max = Math.max(max, stats.total);
+		return { total: stats.total };
+	});
+
+	let requestsChart = processRequestsChart(records, max);
+
+	return {
+		total: stats.total,
+		totalFormatted: formatNum(stats.total),
+		requestsChart,
+	};
+};
+
+const composeTemplate = async (name) => {
+	let [ metadata, stats ] = await Promise.all([ prepareMetadata(name), prepareStats(name) ]);
+
+	return {
+		name: metadata.name,
```

Comment:
Needs handling of long names with `...`: test with e.g. `/og-image/npm/@akqa-frontline/font-config-webpack-plugin`

Reply:
For scoped packages, I think we can go with two lines: first line the scope, in a smaller font (regular text or something between), the second line the name (otherwise the name, which is the more important part, would mostly get cut). If there's not enough space, we can make the description just one line for these. See if you can make this look ok and if not, ping we can ping Patryk for suggestions.

Reply:
Something like this should work. Both scope and name could be pretty long and must be cut but in most cases this should work pretty well

![image](https://user-images.githubusercontent.com/3776001/148933141-d037a666-084b-431b-a55b-e5687e6fe772.png)

Reply:
But we should take in account that og images will be mostly used for preview in messengers/twitter/etc. and elements can't be too small. The regular twitter preview is just about 500px width on desktop

Reply:
> But we should take in account that og images will be mostly used for preview in messengers/twitter/etc. and elements can't be too small. The regular twitter preview is just about 500px width on desktop

For sure, I meant it should be either as the description text or between that size and the previous headline size. Looking at the preview here, I think I'd make the scope a little bigger.

Reply:
UPD with the latest changes to the image